python/pyspark/testing/utils.py

```diff
@@ -209,3 +233,71 @@ def check_error(
         self.assertEqual(
             expected, actual, f"Expected message parameters was '{expected}', got '{actual}'"
         )
+
+
+def assertDataFrameEqual(
+    df: DataFrame, expected: Union[DataFrame, List[Row]], ignore_row_order: bool = True
+):
+    if df is None and expected is None:
+        return True
+    elif df is None or expected is None:
+        return False
+
+    def compare(r1: Row, r2: Row):
+        if r1 is None and r2 is None:
+            return True
+        elif r1 is None or r2 is None:
+            return False
+
+        d1 = r1.asDict()
+        d2 = r2.asDict()
+
+        for key in d1.keys() & d2.keys():
+            if isinstance(d1[key], float) and isinstance(d2[key], float):
+                if abs(d1[key] - d2[key]) > 1e-5:
+                    return False
+            else:
+                if d1[key] != d2[key]:
+                    return False
+        return True
+
+    def assert_schema_equal(
+        df_schema: Optional[Union[AtomicType, StructType, str, List[str], Tuple[str, ...]]],
+        expected_schema: Optional[Union[AtomicType, StructType, str, List[str], Tuple[str, ...]]],
+    ):
```

Comment:
Not that this is a private function in `assertDataFrameEqual`, we can expect the schemas as `StructType` only?